src/wallet/wallet.cpp

```diff
@@ -1462,6 +1522,89 @@ bool CWallet::SetHDChain(const CHDChain &chain, bool memonly) {
     return true;
 }
 
+bool CWallet::SetMnemonicConatiner(const MnemonicConatiner& mnConatiner, bool memonly) {
+    if (!memonly && !CWalletDB(strWalletFile).WriteMnemonic(mnConatiner))
+        throw runtime_error(std::string(__func__) + ": writing chain failed");
+    mnemonicConatiner = mnConatiner;
+    return true;
+}
+
+bool CWallet::EncryptMnemonicConatiner(const CKeyingMaterial& vMasterKeyIn)
+{
+    if (!IsCrypted())
+        return false;
+
+    if (mnemonicConatiner.IsCrypted())
+        return true;
+
+    uint256 id = uint256S(hdChain.masterKeyID.GetHex());
+
+    std::vector<unsigned char> cryptedSeed;
+    if (!EncryptSecret(vMasterKeyIn, mnemonicConatiner.GetSeed(), id, cryptedSeed))
+        return false;
+    SecureVector secureCryptedSeed(cryptedSeed.begin(), cryptedSeed.end());
+    if (!mnemonicConatiner.SetSeed(secureCryptedSeed))
+        return false;
+
+    SecureString mnemonic;
+    if (mnemonicConatiner.GetMnemonic(mnemonic)) {
+        std::vector<unsigned char> cryptedMnemonic;
+        SecureVector vectorMnemonic(mnemonic.begin(), mnemonic.end());
+
+        if ((!mnemonic.empty() && !EncryptSecret(vMasterKeyIn, vectorMnemonic, id, cryptedMnemonic)))
+            return false;
+
+        SecureString secureCryptedMnemonic(cryptedMnemonic.begin(), cryptedMnemonic.end());
+        SecureString secureCryptedPassphrase;
+        if (!mnemonicConatiner.SetMnemonic(secureCryptedMnemonic, secureCryptedPassphrase, false))
+            return false;
+    }
+
+    mnemonicConatiner.SetCrypted(true);
+
+    return true;
+}
+
+bool CWallet::DecryptMnemonicConatiner(MnemonicConatiner& mnConatiner) const
+{
+    if (!IsCrypted())
+        return true;
+
+    if (!mnemonicConatiner.IsCrypted())
+        return false;
+
+    uint256 id = uint256S(hdChain.masterKeyID.GetHex());
+
+    SecureVector seed;
+    SecureVector cryptedSeed = mnemonicConatiner.GetSeed();
+    std::vector<unsigned char> vCryptedSeed(cryptedSeed.begin(), cryptedSeed.end());
+    if (!DecryptSecret(vMasterKey, vCryptedSeed, id, seed))
+        return false;
+
+    mnConatiner = mnemonicConatiner;
+    if (!mnConatiner.SetSeed(seed))
+        return false;
+
+    SecureString cryptedMnemonic;
+
+    if (mnemonicConatiner.GetMnemonic(cryptedMnemonic)) {
+        SecureVector vectorMnemonic;
+
+        std::vector<unsigned char> CryptedMnemonic(cryptedMnemonic.begin(), cryptedMnemonic.end());
+        if (!CryptedMnemonic.empty() && !DecryptSecret(vMasterKey, CryptedMnemonic, id, vectorMnemonic))
+            return false;
+
+        SecureString mnemonic(vectorMnemonic.begin(), vectorMnemonic.end());
+        SecureString passphrase;
+        if (!mnConatiner.SetMnemonic(mnemonic, passphrase, false))
+            return false;
+    }
+
+    mnConatiner.SetCrypted(false);
```

Comment:
Does it remain unencrypted in memory until client quits? Can we use the same time-based mechanism to remove it from memory when the wallet locks on timeout? Or after there is no need in reading mnemonic anymore?

Reply:
It is being deleleted, when going out of scope.